firedrake/slate/slate.py

```diff
@@ -203,16 +213,49 @@ def rank(self):
     def coefficients(self):
         """Returns a tuple of coefficients associated with the tensor."""
 
+    def slate_coefficients(self):
+        return self.coefficients() 
+
     @property
     def coeff_map(self):
         """A map from local coefficient numbers
         to the split global coefficient numbers.
         The split coefficients are defined on the pieces of the originally mixed function spaces.
         """
-        return tuple((n, tuple(range(len(c.split()))))
-                     if isinstance(c, Function) or isinstance(c, Constant)
-                     else (n, (0,))
-                     for n, c in enumerate(self.coefficients()))
+        coeff_map = []
+        orig_to_pos_dict = {}
+        for c in self.slate_coefficients():
+            m = self.slate_coefficients().index(c) 
+            # BlockFunctions for c need special treatment because:
+            # 1) There can be multiple BlockFunctions defined on the same original function
+            # -> The split functions in c might need to be appended to an already exisiting entry in the map
+            # 2) BlockAssembledVectors can appear in any order they like in the Slate expression
+            # -> The split functions need to be appended in the same order into the map
+            if isinstance(c, BlockFunction):  # for block assembled vectors
+                # Did we already add a part of the originial function of this BlockFunction to the map?
+                orig_to_pos = tuple(*filter(lambda item: item[0] == c.orig_function, orig_to_pos_dict.items()))
+                pos = orig_to_pos[1] if orig_to_pos else m
+                # We didn't -> we add normally and keep track that we did in the orig_to_pos_dict
+                # We did -> update already existing entry
+                if pos == m:
+                    split_map = c.indices[0]
+                else:
+                    if c.indices[0]:
+                        split_map = coeff_map[pos][1] + tuple(i for i in c.indices[0] if i not in coeff_map[pos][1])
+                    else:
+                        split_map = coeff_map[pos][1]
+                if pos < m:
+                    # We did -> update already existing entry
+                    coeff_map[pos] = (pos, split_map)
+                else:
+                    # We didn't -> we add normally and keep track that we did in the orig_to_pos_dict
+                    coeff_map += [(pos, split_map)]
+                orig_to_pos_dict[c.orig_function] = pos
+            else:
+                split_map = tuple(range(len(c.split()))) if isinstance(c, Function) or isinstance(c, Constant) else tuple(range(1))
+                coeff_map += [(m, split_map)]
+                orig_to_pos_dict[c] = m
+        return tuple(coeff_map)
```

Comment:
Maybe something this works:
```
    @property
    def coeff_map(self):
        coeff_map = defaultdict(set)
        for c in self.slate_coefficients():
            if isinstance(c, BlockFunction):  # for block assembled vectors
                m = self.coefficients().index(c.orig_function)
                coeff_map[m].update(c.indices[0])
            else:
                m = self.coefficients().index(c)
                split_map = tuple(range(len(c.split()))) if isinstance(c, Function) or isinstance(c, Constant) else tuple(range(1))
                coeff_map[m].update(split_map)
        return tuple((k, tuple(sorted(v)))for k, v in coeff_map.items())
```
`dict` is ordered for `>=3.6`(?), but I don't think it matters here since we now order in `collect_coefficients` using `coeff_map`. We need to check if this works with `initialize_temiminals`, though.